lib/srv/desktop/rdp/rdpclient/src/piv.rs

```diff
@@ -0,0 +1,417 @@
+use crate::errors::invalid_data_error;
+use iso7816::aid::Aid;
+use iso7816::command::instruction::Instruction;
+use iso7816::command::Command;
+use iso7816::response::Status;
+use iso7816_tlv::ber::{Tag, Tlv, Value};
+use openssl::pkey::Private;
+use openssl::rsa::{Padding, Rsa};
+use rdp::model::error::*;
+use std::convert::TryFrom;
+use std::io::{Cursor, Read};
+use uuid::Uuid;
+
+// AID of PIV application, per
+// https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf
+const PIV_AID: Aid = Aid::new_truncatable(
+    &[
+        0xA0, 0x00, 0x00, 0x03, 0x08, 0x00, 0x00, 0x10, 0x00, 0x01, 0x00,
+    ],
+    5, // usually truncates to first 5 bytes
+);
+
+#[derive(Debug)]
+pub struct Card<const S: usize> {
+    chuid: Vec<u8>,
+    piv_auth_cert: Vec<u8>,
+    piv_auth_key: Rsa<Private>,
+    pending_command: Option<Command<S>>,
+    pending_response: Option<Cursor<Vec<u8>>>,
+}
+
+impl<const S: usize> Card<S> {
+    pub fn new(uuid: Uuid, cert_der: &[u8], key_der: &[u8]) -> RdpResult<Self> {
+        Ok(Self {
+            chuid: Self::build_chuid(uuid),
+            piv_auth_cert: Self::build_piv_auth_cert(cert_der),
+            piv_auth_key: Rsa::private_key_from_der(key_der).or_else(|e| {
+                Err(invalid_data_error(&format!(
+                    "failed to parse private key from DER: {:?}",
+                    e
+                )))
+            })?,
```

Comment:
nit: split complex bits like these into a variable with the same name as the field prior to make the code more readable. generally you want to keep it to one line per field or so and initialize a variable for it prior if you can't